### PR TITLE
Rework UCX version

### PR DESCRIPTION
**Pull request by amaslenn**
Main changes are:
 - adding `PATCH_VERSION`: `MAJOR.MINOR.PATCH`
 - switching from `REV_COUNT` to `PATCH_VERSION`
 - redefining `VERSION` as `MAJOR.MINOR.PATCH.MICRO_VERSION` where `MICRO_VERSION` is short sha-1 (7 chars).

### CODE REVIEW

**Comment by shamisp** [comment]
I like it on the conceptual level. Please address the failures.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4188/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4189/ for details (Mellanox internal link).

**Comment by amaslenn** [comment]
@shamisp could you please check your CI? Seems it didn't re-check latest amend commit.

**Comment by boehms** [comment]
ORNL CI fails with the following error:

```
Build rpms
+ ../contrib/buildrpm.sh -s -b
find: paths must precede expression: ucx-1.3.0.b27f99c.tar.gz
Usage: find [-H] [-L] [-P] [-Olevel] [-D help|tree|search|stat|rates|opt|exec] [path...] [expression]
Build step 'Execute shell' marked build as failure
```

File: configure.ac [hunk 1]
```diff
@@ -11,8 +11,8 @@ AC_PREREQ([2.63])
 define([scm_r], esyscmd([sh -c "git rev-parse --short=7 HEAD"]))
```
Comment (shamisp) [inline configure.ac]: First we have to check if git is present  and does not return and error. We have but open for this. Please address this issue as well.

File: src/uct/api/version.h.in [hunk 2]
```diff
@@ -10,9 +10,7 @@
 
 #define UCT_VERNO_MAJOR            @MAJOR_VERSION@
 #define UCT_VERNO_MINOR            @MINOR_VERSION@
-#define UCT_VERNO_REV              "@MICRO_VERSION@"
-#define UCT_VERNO_MICRO            @REV_COUNT@
-#define UCT_VERNO_STRING           "@MAJOR_VERSION@.@MINOR_VERSION@.@REV_COUNT@"
+#define UCT_VERNO_STRING           "@VERSION@"
```
Comment (amaslenn) [inline src/uct/api/version.h.in]: Yes, they will be different.
Currently `ucp_get_version_string()` returns string version similar to what is used in `ucp_get_version()` where there are only 3 numbers: `major`, `minor` and `patch` (previously `rev_count`). In this context it is consistent.

File: configure.ac (hunk 1 above)
Comment (amaslenn) [inline configure.ac]: Without git we can't create correct version of UCX. Is that OK to run build on a machine without git?

File: configure.ac [hunk 3]
```diff
@@ -49,21 +49,17 @@ AC_PREFIX_DEFAULT([ucx_prefix])
...
 
-RPM_RELEASE=1
-BUILD_TS=ts
 MAJOR_VERSION=ucx_ver_major
 MINOR_VERSION=ucx_ver_minor
+PATCH_VERSION=ucx_ver_patch
 MICRO_VERSION=scm_r
-REV_COUNT=revcount
-VERSION=$MAJOR_VERSION.$MINOR_VERSION.$REV_COUNT
+VERSION=$MAJOR_VERSION.$MINOR_VERSION.$PATCH_VERSION.$MICRO_VERSION
```
Comment (amaslenn) [inline configure.ac]: Will fix.

File: configure.ac (hunk 1 above)
Comment (miked-mellanox) [inline configure.ac]: some1 can download tarball and re-run autogen/configure sequence and .git/ will not be available.
maybe we can add check AC_PROG_GIT, smth like https://github.com/bdrewery/pkgng/blob/master/configure.ac#L122

Comment (amaslenn) [inline configure.ac]: OK, got you. I can add this check, no problem.
But what value should I use for `SCM_VERSION` in this case? Empty? `0000000`?

Comment (miked-mellanox) [inline configure.ac]: maybe we should package into tarball VERSION file w/ git hash used during creation of tarball, and if git not found - use it.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4200/ for details (Mellanox internal link).

File: src/uct/api/version.h.in (hunk 2 above)
Comment (yosefe) [inline src/uct/api/version.h.in]: ok

File: configure.ac (hunk 1 above)
Comment (yosefe) [inline configure.ac]: agree with @miked-mellanox and @shamisp :
1. package VERSION into dist tarball
1. recover from "git rev-parse" failure - if .git or /usr/bin/git not present

File: configure.ac [hunk 4]
```diff
@@ -49,21 +49,17 @@ AC_PREFIX_DEFAULT([ucx_prefix])
...
-RPM_RELEASE=1
-BUILD_TS=ts
 MAJOR_VERSION=ucx_ver_major
 MINOR_VERSION=ucx_ver_minor
-MICRO_VERSION=scm_r
-REV_COUNT=revcount
-VERSION=$MAJOR_VERSION.$MINOR_VERSION.$REV_COUNT
+PATCH_VERSION=ucx_ver_patch
+SCM_VERSION=scm_r
+VERSION=$MAJOR_VERSION.$MINOR_VERSION.$PATCH_VERSION.$SCM_VERSION
```
Comment (yosefe) [inline configure.ac]: @shamisp commented yesterday that we do not want the $SCM_VERSION in rpm name.
1. Is this prohibited by Fedora?
1. Is the a better way to add git version info to the rpm file?

File: configure.ac (hunk 1 above)
Comment (amaslenn) [inline configure.ac]: Done.

File: configure.ac (hunk 4 above)
Comment (amaslenn) [inline configure.ac]: 1. It is OK.
2. Please review latest commits.

**Comment by amaslenn** [comment]
7 zeros for revision will be a flag that revision is unknown. It is used for `info` tool only.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4203/ for details (Mellanox internal link).

File: contrib/buildrpm.sh [hunk 5]
```diff
@@ -46,6 +46,15 @@ if [ $opt_tarball -eq 1 ]; then
     make dist
 fi
 
+# Version includes revision, while tarball in Source doesn't have it since
+# it uses GitHub standart name v<Version>.tar.gz, so make:
+# ucx-1.3.0.6a61458.tar.gz --> v1.3.0.tar.gz for rpmbuild
```
Comment (amaslenn) [inline contrib/buildrpm.sh]: Not sure. rpmbuild takes file name from `Source` tag which is `v1.3.0` while autotools create `ucx-1.3.0` tarball (`<name>-<version>`).
So this code is required to pass CI.

File: src/tools/info/build_info.c [hunk 6]
```diff
@@ -16,7 +16,7 @@
 
 void print_version()
 {
-    printf("# UCT version=%s\n", UCT_VERNO_STRING);
+    printf("# UCT version=%s rev.%s\n", UCT_VERNO_STRING, UCT_SCM_VERSION);
```
Comment (amaslenn) [inline src/tools/info/build_info.c]: Will update.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4209/ for details (Mellanox internal link).

**Comment by amaslenn** [comment]
@MattBBaker could you please post ORNL output?

**Comment by amaslenn** [comment]
@MattBBaker BTW, what version of `autoconf`/`aclocal` do you use?

**Comment by boehms** [comment]
@amaslenn Here is the output of the latest build:

```
make[1]: Leaving directory `/var/lib/jenkins/jobs/UCX/workspace/build-tests/ucx-1.3.0/_build'
{ test ! -d "ucx-1.3.0" || { find "ucx-1.3.0" -type d ! -perm -200 -exec chmod u+w {} ';' && rm -fr "ucx-1.3.0"; }; }
===========================================
ucx-1.3.0 archives ready for distribution: 
ucx-1.3.0.tar.gz
===========================================
+ '[' -x /usr/bin/dpkg-buildpackage ']'
+ echo 'Build rpms'
Build rpms
+ ../contrib/buildrpm.sh -s -b
+ rpmbuild -bs '--define=_rpmdir /var/lib/jenkins/jobs/UCX/workspace/build-tests/rpm-dist' '--define=_srcrpmdir /var/lib/jenkins/jobs/UCX/workspace/build-tests/rpm-dist' '--define=_sourcedir /var/lib/jenkins/jobs/UCX/workspace/build-tests' '--define=_specdir /var/lib/jenkins/jobs/UCX/workspace/build-tests' '--define=_builddir /var/lib/jenkins/jobs/UCX/workspace/build-tests' --nodeps --buildroot=/var/lib/jenkins/jobs/UCX/workspace/build-tests/_rpm ucx.spec
Wrote: /var/lib/jenkins/jobs/UCX/workspace/build-tests/rpm-dist/ucx-1.3.0-1.el6.src.rpm
+ rpmbuild -bb '--define=_rpmdir /var/lib/jenkins/jobs/UCX/workspace/build-tests/rpm-dist' '--define=_srcrpmdir /var/lib/jenkins/jobs/UCX/workspace/build-tests/rpm-dist' '--define=_sourcedir /var/lib/jenkins/jobs/UCX/workspace/build-tests' '--define=_specdir /var/lib/jenkins/jobs/UCX/workspace/build-tests' '--define=_builddir /var/lib/jenkins/jobs/UCX/workspace/build-tests' --nodeps --buildroot=/var/lib/jenkins/jobs/UCX/workspace/build-tests/_rpm ucx.spec
Executing(%prep): /bin/sh -e /var/tmp/rpm-tmp.hrXQXJ
+ umask 022
+ cd /var/lib/jenkins/jobs/UCX/workspace/build-tests
+ LANG=C
+ export LANG
+ unset DISPLAY
+ cd /var/lib/jenkins/jobs/UCX/workspace/build-tests
+ rm -rf ucx-1.3.0
+ /usr/bin/gzip -dc /var/lib/jenkins/jobs/UCX/workspace/build-tests/v1.3.0.tar.gz
+ /bin/tar -xf -
+ STATUS=0
+ '[' 0 -ne 0 ']'
+ cd ucx-1.3.0
/var/tmp/rpm-tmp.hrXQXJ: line 38: cd: ucx-1.3.0: No such file or directory
error: Bad exit status from /var/tmp/rpm-tmp.hrXQXJ (%prep)
```
autoconf is 2.6.3 and aclocal is 1.11.1

**Comment by amaslenn** [comment]
@boehms Thanks!
Do you use `test_jenkins.sh` for CI? Or is it something else?

**Comment by boehms** [comment]
@amaslenn We are using a different script.

**Comment by MattBBaker** [comment]
bot:ornl:retest

**Comment by MattBBaker** [comment]
Poked things for a bit, I'm beginning to suspect it was a transient nfs problem. Though I did find that if you run the current buildrpm.sh in a directory that has ucx.spec but no tarball it will create a circular link to ```ucx*.tar.gz```

**Comment by MattBBaker** [comment]
bot:ornl:retest

**Comment by yosefe** [comment]
@shamisp @miked-mellanox can you pls take a look as well?

**Comment by shamisp** [comment]
is the hash still part of rpm name ?

**Comment by amaslenn** [comment]
@shamisp no, rpm name includes only name and `VERSION` (which is `1.3.0`, for example).

**Comment by shamisp** [comment]
👍

**Comment by alinask** [comment]
@shamisp Do you have additional comments for this PR?

**Pull request merged**